helios-client/src/main/java/com/spotify/helios/client/DefaultRequestDispatcher.java

```diff
@@ -77,20 +73,16 @@
 
   private static final Logger log = LoggerFactory.getLogger(DefaultRequestDispatcher.class);
 
-  private static final long RETRY_TIMEOUT_MILLIS = SECONDS.toMillis(60);
   private static final long HTTP_TIMEOUT_MILLIS = SECONDS.toMillis(10);
-  private static final List<String> VALID_PROTOCOLS = ImmutableList.of("http", "https");
-  private static final String VALID_PROTOCOLS_STR =
-      String.format("[%s]", Joiner.on("|").join(VALID_PROTOCOLS));
 
-  private final Supplier<List<URI>> endpointSupplier;
+  private final Iterator<Endpoint> endpointIterator;
   private final ListeningExecutorService executorService;
   private final String user;
 
-  public DefaultRequestDispatcher(final Supplier<List<URI>> endpointSupplier,
-                                  final String user,
-                                  final ListeningExecutorService executorService) {
-    this.endpointSupplier = endpointSupplier;
+  DefaultRequestDispatcher(final Supplier<List<Endpoint>> endpointSupplier,
+                           final String user,
+                           final ListeningExecutorService executorService) {
+    endpointIterator = EndpointIterator.of(endpointSupplier.get());
```

Comment:
if we only ever call `Supplier.get()` once, I think you could just simplify and have the constructor parameter be the actual list / iterable